### PR TITLE
Enables cache and parallel builds for tests

### LINKED ISSUE #211: use cacher to reduce Travis CI build time
Travis CI has a way to [cache dependencies](https://docs.travis-ci.com/user/caching/); there is a way to use the feature for [python](https://docs.travis-ci.com/user/caching/#pip-cache). This feature along with virtenv should reduce the build time tremendously.

### PR DESCRIPTION
**Pull request by bhuvangupta008**
## What? Why?
Fix #211  . (Add Reference to the issue if any)

Changes proposed in this pull request:
- a Enabled Cache for pip packages
- b Build matrix(Unit test, Functional test, Integration test)

## Notify reviewers
@kaushik-rohit
@prasadtalasila

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -23,16 +25,13 @@ install:
...
-  - bash ext/doc_auto_deploy.sh
+  bash $TEST_SUITE
 
 after_success:
-  - coverage report -m
-  - codecov
-  - cd test
-  #- mprof clean && mprof run python -m unittest -v tests && cd ../profiler && python memory_profiler.py && echo "FOR UPDATED MEMORY CONSUMPTION, see image:" && curl -F "clbin=@memory_consumption.png" https://clbin.com
-  #- sfood ../lib/analysis ../lib/analysis/* ../ext ../lib/in_out/* ../lib/config.py ../lib/nickTracker.py ../lib/util.py ../lib/vis.py ../sample.py | sfood-graph | dot -Tpng -Gsize=9,15\! -Gdpi=300 -o call-graph-dependency.png && echo "FOR UPDATED DEPENDENCY, see image:" && curl -F "clbin=@call-graph-dependency.png" https://clbin.com
+  - bash ext/doc_auto_deploy.sh
```
Comment (prasadtalasila) [inline .travis.yml]: With the current structure, the doc_auto_deploy.sh is running in all the runs of the build matrix. A better solution is to mix the build stages with build matrix. Please see [example here](https://docs.travis-ci.com/user/build-stages/matrix-expansion/). The `bash ext/doc_auto_deploy.sh` command can be pushed to deploy stage. The default stage would be a test stage (this fact is also exploited in the example given).

**Review by prasadtalasila — CHANGES_REQUESTED** [review]
@bhuvangupta008 please see the suggestions.

File: .travis.yml (hunk 1 above)
Comment (bhuvangupta008) [reply .travis.yml]: @prasadtalasila Sir, if we do the same as inside example and deploy is pushed to a separate stage, then a new VM is created and all the dependency installations are completed. i've pushed the commit already. you can have a view of the travis build. If this is fine, we can go ahead and merge but this deployment then got its's own 4.5 minute setup time thus pushing the total time to 15 mins.

**Comment by prasadtalasila** [comment]
@bhuvangupta008 In that case, lets keep the docs deployment in unit tests. That should reduce the build time to 8 minutes.

After all of this work, the build time goes down by only 1 minute for now. But as we add more functional and integration tests, the benefits become apparent.

**Comment by bhuvangupta008** [comment]
@prasadtalasila, done sir.

**Review by prasadtalasila — APPROVED** [review]

**Pull request merged**